### PR TITLE
IDC: rename the sync_idc_optin method, filter, and constant to should_handle_idc

### PR DESCRIPTION
**Pull request by kbrown9**
#### Changes proposed in this Pull Request:
In the future, IDC detection and mitigation will no longer be dependent on Sync communications. Let's remove references to `sync` from the `sync_idc_optin` method, filter, and constant to avoid confusion.

* Deprecate the `sync_idc_optin()` method.
* Add the new `should_handle_idc()` method, which replaces `sync_idc_optin()`.
* Add a new constant, `JETPACK_SHOULD_HANDLE_IDC`. Check the new constant before checking the legacy `JETPACK_SYNC_IDC_OPTIN` constant.
* Add a new `jetpack_should_handle_idc` filter.
* Deprecate the `jetpack_sync_idc_optin` filter.
* Apply the legacy filter in a callback hooked to the new `jetpack_should_handle_idc` filter.
* Update the unit tests to use the new method, filter, and constant.
* Add tests for the legacy constant and filter.

#### Jetpack product discussion
* p9dueE-3oW-p2

#### Does this pull request change what data or activity we track or use?
* no

#### Testing instructions:

Note: I used the Code Snippets plugin to add the snippet in step 2a and to set the filters and constants in the later tests.

##### Verify that IDC detection works
1. Install this branch of Jetpack. Activate and connect to WPCOM.
2. Cause an IDC:
   a. You can use something like this snippet to cause an IDC.
```
        add_filter( 'jetpack_sync_site_url', function ( $value ) {
	    if( defined('XMLRPC_REQUEST') && XMLRPC_REQUEST ) {
	        return $value;
	    }
	    return 'http://example.org/site';
        } );
```
3. Navigate to a page on wp-admin and verify that the IDC notice is displayed.
4. Navigate to wp-admin -> Tools -> Site Health. Click the Info tab at the top, then expand the Jetpack section. Scroll down to the bottom and verify that the value of `IDC Opt-in` is 1. Also verify that the IDC Error Option is populated with the remote and wpcom urls and the error code.

##### Verify that IDC detection can be turned off using the filters and constants
In the following tests, we'll change the siteurl by activating/deactivating the snippet in Step 2. The IDC will not be detected because we'll use the filters and constants to turn off IDC handling.

You'll need your sandbox for checking the value of the wpcom side siteurl value. (You don't need to point your site to your sandbox, you'll just use it to check values.)

###### Test the new filter, `jetpack_should_handle_idc`
6. Set up a hook to cause `jetpack_should_handle_idc` to return false:
` add_filter('jetpack_should_handle_idc', '__return_false');`
7. Navigate to wp-admin -> Tools -> Site Health. Click the Info tab at the top, then expand the Jetpack section. Scroll down to the bottom and verify that the value of `IDC Opt-in` is empty.
8. Cause the siteurl to change by activating the snippet in steps 2 and causing a sync event. For example, edit a post.
9. Since the `jetpack_should_handle_idc` filter returns false, the IDC will not be detected. Using your sandbox, check the value of the blog's `siteurl` option. It should have changed to the siteurl value returned in step 2a.

###### Test the legacy filter, `jetpack_sync_idc_optin`
10. Set up a hook to cause `jetpack_sync_idc_optin` to return false:
` add_filter('jetpack_sync_idc_optin', '__return_false');`
11. Navigate to wp-admin -> Tools -> Site Health. Click the Info tab at the top, then expand the Jetpack section. Scroll down to the bottom and verify that the value of `IDC Opt-in` is false.
12. Cause the siteurl to change by deactivating the snippet in steps 2 and causing a sync event. For example, edit a post.
13. Since the `jetpack_sync_idc_optin` filter returns false, the IDC will not be detected. Using your sandbox, check the value of the blog's `siteurl` option. It should have changed to the original siteurl value.

###### Test the new constant, `JETPACK_SHOULD_HANDLE_IDC`
14. Define the `JETPACK_SHOULD_HANDLE_IDC` constant as false:
` define('JETPACK_SHOULD_HANDLE_IDC', false);`
15. Navigate to wp-admin -> Tools -> Site Health. Click the Info tab at the top, then expand the Jetpack section. Scroll down to the bottom and verify that the value of `IDC Opt-in` is false.
16. Cause the siteurl to change by activating the snippet in steps 2 and causing a sync event. For example, edit a post.
17. Since the `JETPACK_SHOULD_HANDLE_IDC` constant is false, the IDC will not be detected. Using your sandbox, check the value of the blog's `siteurl` option. It should have changed to the siteurl value returned in step 2a.

###### Test the legacy constant, `JETPACK_SYNC_IDC_OPTIN`
18. Define the `JETPACK_SYNC_IDC_OPTIN` constant as false:
` define('JETPACK_SYNC_IDC_OPTIN', false);`
19. Navigate to wp-admin -> Tools -> Site Health. Click the Info tab at the top, then expand the Jetpack section. Scroll down to the bottom and verify that the value of `IDC Opt-in` is false.
20. Cause the siteurl to change by deactivating the snippet in steps 2 and causing a sync event. For example, edit a post.
21. Since the JETPACK_SYNC_IDC_OPTIN` constant is false, the IDC will not be detected. Using your sandbox, check the value of the blog's `siteurl` option. It should have changed to the original siteurl value.

##### Check for errors and deprecation notices
22. Verify that no errors or deprecation notices were generated in your site's debug.log file. (Note that the test that uses the `jetpack_sync_idc_optin` filter should generated deprecation notices.)

### CODE REVIEW

File: projects/packages/identity-crisis/src/class-identity-crisis.php [hunk 1]
```diff
@@ -432,14 +432,59 @@ public static function get_sync_error_idc_option( $response = array() ) {
...
+	 *
+	 * This method uses both the current jetpack_should_handle_idc filter and constant and the
+	 * legacy jetpack_sync_idc_optin filter and constant to determine whether an IDC should be
+	 * handled.
+	 *
+	 * @return bool
+	 * @since $$next-version$$
+	 */
+	public static function should_handle_idc() {
+		if ( Constants::is_defined( 'JETPACK_SHOULD_HANDLE_IDC' ) ) {
```
Comment (kbrown9) [inline projects/packages/identity-crisis/src/class-identity-crisis.php:455]: I'm not sure if this new constant is really necessary. I added it as a replacement for the existing `JETPACK_SYNC_IDC_OPTIN` constant, but it seems like the filter (see below) should be sufficient.

Comment (zinigor) [reply projects/packages/identity-crisis/src/class-identity-crisis.php:455]: I guess a constant makes it easier to disable IDC if needed, without having to define a custom filter.

**Review by zinigor — COMMENTED** [review]
I haven't tested this PR yet, but the code looks good to me. I'm planning to return to this on Monday, but I would be OK to merge it since we plan to work with IDC a lot in the near future.

File: projects/packages/identity-crisis/src/class-identity-crisis.php (hunk 1 above)
Comment (mdbitz) [reply projects/packages/identity-crisis/src/class-identity-crisis.php:455]: I like the versatility of both a constant and a filter. However looking at the code wow its complicated to deprecate all the things. Sorry for the original suggestion of the name change

**Review by mdbitz — APPROVED** [review]
Excellent addition of Unit Tests with this change. I tested locally and am seeing the filter/constant apply  properly to the flows.

**Pull request merged**